### PR TITLE
Eden tester

### PR DESCRIPTION
**Pull request by tbfleming**

### CODE REVIEW

File: libraries/eosiolib/tester/wasi_polyfill/__wasi_fd_fdstat_get.cpp [hunk 1]
```diff
@@ -0,0 +1,11 @@
+#include <wasi/api.h>
+
+extern "C" __wasi_errno_t __wasi_fd_fdstat_get(__wasi_fd_t fd, __wasi_fdstat_t* stat)
+    __attribute__((__import_module__("wasi_snapshot_preview1"), __import_name__("fd_fdstat_get")))
+{
+   stat->fs_filetype = __WASI_FILETYPE_CHARACTER_DEVICE;
+   stat->fs_flags = __WASI_FDFLAGS_APPEND;
+   stat->fs_rights_base = __WASI_RIGHTS_FD_READ | __WASI_RIGHTS_FD_WRITE;
```
Comment (swatanabe) [inline libraries/eosiolib/tester/wasi_polyfill/__wasi_fd_fdstat_get.cpp:8]: Are there and readable file descriptors?

File: programs/eden-tester/main.cpp [hunk 2]
```diff
@@ -0,0 +1,1141 @@
...
+         {
+            get_argv_buf() = ch;
+            ++argv_buf;
+         }
+         get_argv_buf() = 0;
+         ++argv_buf;
+      }
+   };
+
+   int32_t clock_gettime(int32_t id, void* data)
```
Comment (swatanabe) [inline programs/eden-tester/main.cpp:687]: This should probably be adjusted to implement the WASI interface.

File: programs/eden-tester/main.cpp [hunk 3]
```diff
@@ -0,0 +1,1141 @@
...
+      }
+      int32_t sec = result.count() / 1000000000;
+      int32_t nsec = result.count() % 1000000000;
+      fc::datastream<char*> ds((char*)data, 8);
+      fc::raw::pack(ds, sec);
+      fc::raw::pack(ds, nsec);
+      return 0;
+   }
+
+   int32_t open_file(span<const char> filename, span<const char> mode)
```
Comment (swatanabe) [inline programs/eden-tester/main.cpp:711]: The file handling here should be integrated with WASI.

File: .gitmodules [hunk 4]
```diff
@@ -1,3 +1,12 @@
 [submodule "external/rapidjson"]
 	path = external/rapidjson
 	url = https://github.com/Tencent/rapidjson.git
+[submodule "external/eos"]
+	path = external/eos
+	url = https://github.com/eoscommunity/eos.git
```
[the remark this answers is not in the document]
Comment (tbfleming) [reply .gitmodules:6]: This keeps the intrinsic implementations and the wrappers together in the same repo until they stabilize. This problem will hopefully be solved after we reorg everything into a single repo.

File: external/CMakeLists.txt [hunk 5]
```diff
@@ -1,2 +1,100 @@
...
+
+if(IS_NATIVE)
+    ExternalProject_Add(eos
+        SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/eos
+        BINARY_DIR eos
+        BUILD_ALWAYS 1
+        TEST_EXCLUDE_FROM_MAIN 1
+        EXCLUDE_FROM_ALL 1
+        BUILD_COMMAND cmake --build . -t eosio_chain -j ${NUM_PROCS}
+        INSTALL_COMMAND sh -c "mkdir -p ${ROOT_BINARY_DIR}/eos && cp -p `find . -name *.a` ${ROOT_BINARY_DIR}/eos"
```
[the remark this answers is not in the document]
Comment (tbfleming) [reply external/CMakeLists.txt]: fixed

File: libraries/eosiolib/tester/tester.cpp [hunk 6]
```diff
@@ -657,7 +643,7 @@ uint32_t eosio::test_rodeos::sync_blocks()
    return n;
 }
 
-[[nodiscard]] eosio::transaction_trace eosio::test_rodeos::push_transaction(
+[[nodiscard]] eosio::transaction_trace eosio::test_rodeos::tester_push_transaction(
```
[the remark this answers is not in the document]
Comment (tbfleming) [reply libraries/eosiolib/tester/tester.cpp]: No, fixed

File: programs/eden-tester/main.cpp (hunk 2 above)
Comment (tbfleming) [reply programs/eden-tester/main.cpp]: done

File: programs/eden-tester/main.cpp (hunk 3 above)
Comment (tbfleming) [reply programs/eden-tester/main.cpp]: done

File: libraries/eosiolib/tester/wasi_polyfill/__wasi_fd_fdstat_get.cpp (hunk 1 above)
Comment (tbfleming) [reply libraries/eosiolib/tester/wasi_polyfill/__wasi_fd_fdstat_get.cpp]: Before: ish. Even though stdin wasn't really hooked up, this allowed musl to not break during initialization. Having extra rights on stdin, stdout, and stderr didn't hurt it.

Now: replaced with native-side implementation which deals with files in addition to stdin, stdout, and stderr

File: libraries/eosiolib/tester/wasi_polyfill/__wasi_fd_prestat_dir_name.cpp [hunk 7]
```diff
@@ -6,8 +6,12 @@ extern "C" __wasi_errno_t __wasi_fd_prestat_dir_name(__wasi_fd_t fd,
     __attribute__((__import_module__("wasi_snapshot_preview1"),
                    __import_name__("fd_prestat_dir_name")))
 {
-   [[clang::import_name("prints")]] void prints(const char*);
-   [[clang::import_name("tester_abort"), noreturn]] void tester_abort();
-   prints("__wasi_fd_prestat_dir_name not implemented");
-   tester_abort();
+   // root dir
+   if (fd == 3)
```
[the remark this answers is not in the document]
Comment (tbfleming) [reply libraries/eosiolib/tester/wasi_polyfill/__wasi_fd_prestat_dir_name.cpp]: fixed. Used `polyfill_root_dir_fd` to emphasize it's not part of wasi's spec, which allows an arbitrary set of preopened dirs, sometimes excluding root.

File: programs/eden-tester/main.cpp [hunk 8]
```diff
@@ -698,47 +717,160 @@ struct callbacks
       }
       else
       {
-         return -1;
+         return EINVAL;
```
[the remark this answers is not in the document]
Comment (tbfleming) [reply programs/eden-tester/main.cpp]: fixed

**Review by swatanabe — APPROVED** [review]

**Pull request merged**